x-pack/plugins/stack_connectors/public/connector_types/security/tines/tines_params.test.tsx

```diff
@@ -0,0 +1,537 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import React from 'react';
+import { mountWithIntl } from '@kbn/test-jest-helpers';
+import { MockCodeEditor } from '@kbn/triggers-actions-ui-plugin/public/application/code_editor.mock';
+import type { UseSubActionParams } from '@kbn/triggers-actions-ui-plugin/public/application/hooks/use_sub_action';
+import TinesParamsFields from './tines_params';
+
+const kibanaReactPath = '@kbn/kibana-react-plugin/public';
+const triggersActionsPath = '@kbn/triggers-actions-ui-plugin/public';
+interface Result {
+  isLoading: boolean;
+  response: Record<string, unknown>;
+  error: null | Error;
+}
+const mockUseSubActionStories = jest.fn<Result, [UseSubActionParams<unknown>]>(() => ({
+  isLoading: false,
+  response: { stories: [story], incompleteResponse: false },
+  error: null,
+}));
+const mockUseSubActionWebhooks = jest.fn<Result, [UseSubActionParams<unknown>]>(() => ({
+  isLoading: false,
+  response: { webhooks: [webhook], incompleteResponse: false },
+  error: null,
+}));
+const mockUseSubAction = jest.fn<Result, [UseSubActionParams<unknown>]>((params) =>
+  params.subAction === 'stories'
+    ? mockUseSubActionStories(params)
+    : mockUseSubActionWebhooks(params)
+);
+
+const mockToasts = { danger: jest.fn(), warning: jest.fn() };
+jest.mock(triggersActionsPath, () => {
+  const original = jest.requireActual(triggersActionsPath);
+  return {
+    ...original,
+    useSubAction: (params: UseSubActionParams<unknown>) => mockUseSubAction(params),
```

Comment:
This is a unit test for the form, I don't think we should test the `useSubAction` hook along with it. The `useSubAction` hook has its own unit test, and there's also an integration test that checks that all the UI components used in this form work properly together.

We should not have to come and do changes to this unit test unless there's some change in the component that it is testing. Otherwise, it would mean that this is not a unit but an integration test, and we already have one.

I mean, if we don't mock the `useSubAction` in the unit tests of the components that use it, a change in the internal logic of this hook may cause all those tests to fail, even if nothing changed from the outside and everything is working fine, introducing false negatives and making us go and fix all those tests unnecessarily.
This is not how a unit test should work. they should have only one responsibility, test only one thing, and we should avoid coupling them with other things.

Reply:
We do not test the `useSubAction` by not mocking it. We test how our component interacts with the hook. Imagine having a hook with one `useEffect`. By mocking the hook you do not really test how your component behaves with the extra `useEffect` because you mocked it. Also, by mocking it we do not test how our component behaves with async operations like fetching etc.

> I mean, if we don't mock the useSubAction in the unit tests of the components that use it, a change in the internal logic of this hook may cause all those tests to fail, even if nothing changed from the outside and everything is working fine, introducing false negatives and making us go and fix all those tests unnecessarily.

This is more of a reason to not mock the hook. If we do a change in the hook and the tests of the component break then it means that we introduce a bug in the hook. By mocking it you will not be aware of that. A change in the hook should pass the tests either way.

>This is not how a unit test should work. they should have only one responsibility, test only one thing, and we should avoid coupling them with other things.?

This is correct. Not mocking the hooks does not mean that you turn a unit test into an integration one. It is still a unit test with only one responsibility, testing the component. And the component is coupled with all hooks it uses.

I know we tend to mock the hooks a lot, especially the ones that fetch, but I think we should stop doing it and mock only what is necessary.  Nevertheless, if you prefer to mock it I am fine. I just wanted to mention it. Some interesting articles about the topic: https://kentcdodds.com/blog/stop-mocking-fetch, https://tkdodo.eu/blog/testing-react-query#mocking-network-requests